modules/hostdb/entries.go

```diff
@@ -11,39 +9,43 @@ import (
 // chosen according to the need at hand. This might also require having a bunch
 // of different weights at each node in the tree.
 
-// host.Weight() determines the weight of a specific host, which is:
+// entryWeight determines the weight of a specific host, which is:
 //
 //		Freeze * Collateral / square(Price).
 //
 // Freeze has to be linear, because any non-linear freeze will invite sybil
 // attacks.
 //
-// For now, Collateral is also linear because an increased burn means
-// increased risk for the host (Freeze on the other hand has no risk). It
-// might be better to make burn grow sublinearly, such as taking
-// sqrt(Collateral) or burn^(4/5).
+// For now, collateral is also linear because an increased collateral means
+// increased risk for the host. (Freeze on the other hand has no risk.) It
+// might be better to make collateral grow sublinearly, such as taking
+// sqrt(collateral) or collateral^(4/5).
 //
 // We take the square of the price to heavily emphasize hosts that have a low
 // price. This is also a bit simplistic however, because we're not sure what
 // the host might be charging for bandwidth.
-func entryWeight(entry modules.HostEntry) consensus.Currency {
+func entryWeight(entry modules.HostEntry) (weight consensus.Currency) {
 	// Catch a divide by 0 error, and let all hosts have at least some weight.
 	//
 	// TODO: Perhaps there's a better way to do this.
-	if entry.Price == 0 {
-		entry.Price = 1
+	if entry.Price.Sign() == 0 {
+		entry.Price = consensus.NewCurrency(1)
 	}
-	if entry.Collateral == 0 {
-		entry.Collateral = 1
+	if entry.Collateral.Sign() == 0 {
+		entry.Collateral = consensus.NewCurrency(1)
 	}
-	if entry.Freeze == 0 {
-		entry.Freeze = 1
+	if entry.Freeze.Sign() == 0 {
+		entry.Freeze = consensus.NewCurrency(1)
 	}
 
-	adjustedCollateral := float64(entry.Collateral)
-	adjustedFreeze := float64(entry.Freeze)
-	adjustedPrice := math.Sqrt(float64(entry.Price))
+	// weight := entry.Freeze * entry.Collateral / sqrt(entry.Price)
+	weight.Add(entry.Freeze)
+	// mathematically, overflow can only occur here
+	err := weight.Mul(entry.Collateral)
+	if err != nil {
+		// TODO: ???
```

Comment:
true. So the maximum possible weight is 2^64 * 2^128 * 2^128 / 1 = 2^320, which is 40 bytes. It wouldn't be so bad to just use `*big.Int`s throughout the weighted list. Solves a lot of the overflow problems, really.